### PR TITLE
Download image

### PR DESCRIPTION
**Pull request by alexandershpak**

### CODE REVIEW

File: db/scripts/V7__add_image_snapshot.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE element_instance ADD image_snapshot	TEXT;
+ALTER TABLE iofog_change_tracking ADD `is_image_snapshot` BIGINT;
+
+DELETE FROM iofog_change_tracking
```
Comment (alexandershpak) [inline db/scripts/V7__add_image_snapshot.sql]: add :

File: src/server/controllers/api/imageSnapshotController.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        async.apply(UserService.getUser, userProps, params),
+        async.apply(ElementInstanceService.getElementInstance, elementInstanceProps),
+        async.apply(ChangeTrackingService.updateChangeTracking, changeTrackingProps)
+
+    ], function(err, result) {
+        AppUtils.sendResponse(res, err, 'id', params.bodyParams.elementId, result);
+    });
+};
+
+/***** instance Image Snapshot of Element Instance EndPoint (Get: /api/v2/instance/imageSnapshotPut/id/:ID/token/:Token *****/
```
Comment (alexandershpak) [inline src/server/controllers/api/imageSnapshotController.js]: correct comment

File: src/server/controllers/api/instanceChangesController.js [hunk 3]
```diff
@@ -95,20 +96,24 @@ const processChangeTrackingChanges = function(params, callback) {
...
     params.changes = changes;
     callback (null, params);
   }else{
     callback('Error', 'Error: Cannot find changeTracking data of current iofog instance.')
   }
 };
 
-const updateChangeTracking = function (params, callback) {
-    var changeTrackingProps = {
+const updateChangeTrackingReboot = function (params, callback) {
```
Comment (alexandershpak) [inline src/server/controllers/api/instanceChangesController.js]: look better

File: src/server/controllers/api/imageSnapshotController.js (hunk 2 above)
Comment (alexandershpak) [reply src/server/controllers/api/imageSnapshotController.js]: fix

File: src/server/controllers/api/imageSnapshotController.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                let stat = fs.statSync(_path);
+                let fileSize = stat.size;
+                res.writeHead(200, {
+                    'Content-Length': fileSize,
+                    "Content-Type" : mimetype
+                });
+                fs.createReadStream(params.elementInstance.imageSnapshot).pipe(res);
+                fs.unlink(params.elementInstance.imageSnapshot, (err) => {
+                    if (err) throw err;
+                    console.log('successfully deleted ' + params.elementInstance.imageSnapshot);
```
Comment (elukashick) [inline src/server/controllers/api/imageSnapshotController.js]: better to use logger.info

File: db/scripts/V5__add_image_snapshot.sql [hunk 5]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE element_instance ADD image_snapshot	TEXT;
+ALTER TABLE iofog_change_tracking ADD `is_image_snapshot` BIGINT;
```
Comment (elukashick) [inline db/scripts/V5__add_image_snapshot.sql]: better to use TINYINT(1)

Comment (alexandershpak) [reply db/scripts/V5__add_image_snapshot.sql]: change `is_image_snapshot` to `image_snapshot`

File: src/server/services/elementInstanceService.js [hunk 6]
```diff
@@ -247,6 +255,14 @@ if (params.bodyParams.instanceId) {
     .then(AppUtils.onUpdate.bind(null, params, "Unable to update 'iofog_uuid' field for Element Instance", callback));
 }
 
+const updateElemInstanceByFogUuIdWithChanges = function(props, params, callback) {
+ let fogInstanceId = AppUtils.getProperty(params, props.fogInstanceId);
```
[the remark this answers is not in the document]
Comment (alexandershpak) [reply src/server/services/elementInstanceService.js]: fix

File: src/server/controllers/api/imageSnapshotController.js (hunk 4 above)
Comment (alexandershpak) [reply src/server/controllers/api/imageSnapshotController.js]: fix

**Pull request merged**